### PR TITLE
Optimise builds

### PR DESCRIPTION
**Pull request by tobiasge**
## New Behavior

- More builds are done in parallel
- Builds are only run when necessary (changes in Netbox, Netboc-docker, Python image)

## Contrast to Current Behavior

- See above

## Discussion: Benefits and Drawbacks
When following the "latest" or "v2.7" tags in Kubernetes or Openshift the containers were deployed everytime the GH Action ran. From now on the images are build only when something changed. So the automatic deployment is only triggered when really needed. 
The image count was also reduced. Changes to the master branch in Netbox are only made when a new release is ready. So we don't need to build extra images for the latest tag. This tag now follows the newest version of Netbox.

## Changes to the Wiki

- "Build" page: Adapt the list of build scripts

## Proposed Release Note Entry

- Maintenance: Improved build system

## Double Check
* [x] I have read the comments and followed the PR template.
* [x] I have explained my PR according to the information in the comments.
* [x] My PR targets the `develop` branch.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-ARG FROM=python:3.7-alpine
+ARG FROM
```
Comment (cimnine) [inline Dockerfile:1]: Why did you remove the default?

File: build.sh [hunk 2]
```diff
@@ -217,22 +224,68 @@ for DOCKER_TARGET in "${DOCKER_TARGETS[@]}"; do
...
+    ###
+    # Load information from registry (only for docker.io)
+    SHOULD_BUILD="false"
+    BUILD_REASON=""
+    if [ -z "${GH_ACTION}" ]; then
+      # Asuming non Github builds should always proceed
+      SHOULD_BUILD="true"
+      BUILD_REASON="${BUILD_REASON} interactive"
+    fi
+    if [ "$DOCKER_REGISTRY" = "docker.io" ] && [ "$SHOULD_BUILD" = "false" ]; then
```
Comment (cimnine) [inline build.sh]:
```suggestion
    elsif [ "$DOCKER_REGISTRY" = "docker.io" ]; then
```

**Comment by cimnine** [comment]
When I read through this, I start to think we should rewrite the build script in something else than bash, Python or so 🙈

**Comment by tobiasge** [comment]
I think for now Bash is OK, but if we want to add more functionality we should do that.

**Comment by tobiasge** [comment]
The builds in my [repository](https://github.com/tobiasge/netbox-docker/actions) with the new script have been stable for a while now. I think we can merge this.

File: build.sh [hunk 3]
```diff
@@ -269,6 +323,10 @@ for DOCKER_TARGET in "${DOCKER_TARGETS[@]}"; do
         --label "NETBOX_GIT_URL=${NETBOX_GIT_URL}"
       )
     fi
+    if [ -n "${BUILD_REASON}" ]; then
+      BUILD_REASON=$(sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//' <<< "$BUILD_REASON")
+      DOCKER_BUILD_ARGS+=( --label "BUILD_REASON=${BUILD_REASON}" )
```
Comment (cimnine) [inline build.sh:327]: What is your idea that you add this as a label? Wouldn't it be enough to just print it to the console?

**Review by cimnine — COMMENTED** [review]
Did you run [shellcheck](https://www.shellcheck.net/) against shell files?

**Comment by tobiasge** [comment]
> Did you run [shellcheck](https://www.shellcheck.net/) against shell files?

I did. But I found just found one more `shellcheck` problem. Will commit a fix to this branch.

File: Dockerfile (hunk 1 above)
Comment (tobiasge) [reply Dockerfile:1]: To check if the source image changed, we have to know what it is. So I moved the default to the `build.sh` script.

File: build.sh (hunk 3 above)
Comment (tobiasge) [reply build.sh:327]: I thought it would be use full to know why a particular image was build without searching through the build logs.

**Review by cimnine — APPROVED** [review]
I still don't understand the whole file, but since it seems working in your branch, let's go for it.

**Comment by cimnine** [comment]
I believe we must update the repository settings shortly after this is merged to specify which of the new jobs are required to run and pass in order for a commit to get 'green'.
